gdal/port/cpl_conv.h

```diff
@@ -358,9 +363,34 @@ class CPL_DLL CPLConfigOptionSetter
     char *m_pszOldValue;
     bool m_bRestoreOldValue;
 };
+}
 
 #endif /* def __cplusplus */
 //! @endcond
 
+#if defined(__cplusplus) && !defined(CPL_SUPRESS_CPLUSPLUS)
+
+extern "C++"
+{
+namespace cpl
+{
+    /** Use cpl::down_cast<Derived*>(pointer_to_base) as equivalent of
+     * static_cast<Derived*>(pointer_to_base) with safe checking in debug
+     * mode.
+     * 
+     * Only works if no virtual inheritance is involved.
+     * 
+     * @param f pointer to a base class
+     * @return pointer to a derived class
+     */
+    template<typename To, typename From> inline To down_cast(From* f)
+    {
+        CPLAssert(f == nullptr || dynamic_cast<To>(f) != nullptr);
```

Comment:
I would suggest doing a static_assert with std::is_base_of first.

Reply:
By nature of the test this is not something that can evaluated at compilation time, but only at runtime. The purpose is, for debug builds, to really ensure that the down_cast was legit; This is inspirted from the https://github.com/googlei18n/google-input-tools/blob/master/client/base/casts.h you mentionned in https://trac.osgeo.org/gdal/ticket/6720

Reply:
You can test the To is a base class of From with a static_assert.   I'm looking at that in code that I don't have permission to share.

Reply:
"You can test the To is a base class of From with a static_assert": you probably meant the reverse ?
But this isn't needed, the compiler will already check that static_assert<> is used to do a potentially legit conversion

For example, the following doesn't compile
```
struct Base
{
    virtual ~Base() {}
};
struct Unrelated
{
};

void foo()
{
    Base b;
    static_cast<Unrelated*>(&b);
}
```

```
test.cpp: In function ‘void foo()’:
test.cpp:12:31: error: invalid static_cast from type ‘Base*’ to type ‘Unrelated*’
     static_cast<Unrelated*>(&b);
```

Reply:
I got permission to share.  Here is what I am suggesting.  I think it is helpful (and cheap) in a few other cases.

```
template<typename To, typename From>
inline To down_cast(From& f) {
{
    static_assert(
        std::is_lvalue_reference<To>::value, "target type not a reference");
    static_assert(
        (std::is_base_of<From, typename std::remove_reference<To>::type>::value),
        "target type not derived from source type");
```

and

```
template<typename To, typename From>
inline To down_cast(From* f) {
  static_assert(
      (std::is_base_of<From, typename std::remove_pointer<To>::type>::value),
      "target type not derived from source type");
```